pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -2190,6 +2193,77 @@ protected void handleGetLastMessageId(CommandGetLastMessageId getLastMessageId)
         }
     }
 
+    private boolean isValidEntry(MessageMetadata metadata, PersistentTopic persistentTopic) {
+        if (metadata.hasMarkerType()) {
+            return false;
+        }
+        if (metadata.hasTxnidMostBits() && metadata.hasTxnidLeastBits() && persistentTopic.isTxnAborted(
+                new TxnID(metadata.getTxnidMostBits(), metadata.getTxnidLeastBits()), null)) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Get the metadata and position of the largest batch index of the last valid entry.
+     * some entry need to be filtered, such as marker message.
+     * @param entryMetaDataFuture
+     * @param lastPosition
+     * @param ml
+     */
+    private void readLastValidEntry(CompletableFuture<MessageMetadata> entryMetaDataFuture,
```

Comment:
This will search forward one by one until a valid position is found, not sure about the performance.

Reply:
If there is an aborted txn with lots of message to be discarded, the search may be time consuming. But time-consuming is better than stucking, right?